smithy-model/src/main/java/software/amazon/smithy/model/node/NumberNode.java

```diff
@@ -94,6 +95,38 @@ public Optional<NumberNode> asNumberNode() {
         return Optional.of(this);
     }
 
+    /**
+     * Returns true if the value of the number contained in the number node is zero,
+     * accounting for float, double, bigInteger, bigDecimal, and other numeric types
+     * (e.g., 0, 0.0, 0.0000000, etc).
+     *
+     * @return Returns true if set to zero.
+     */
+    public boolean isZero() {
+        // Do a cheap test based on the serialized value of the number first.
+        if (toString().equals("0")) {
+            return true;
+        } else if (value instanceof BigDecimal) {
+            return value.equals(BigDecimal.ZERO);
+        } else if (value instanceof BigInteger) {
+            return value.equals(BigInteger.ZERO);
+        } else if (value instanceof Float) {
+            // Using Math.signum to avoid epsilon issues.
+            return Math.signum((Float) value) == 0;
```

Comment:
I'm unsure if we want isZero to return true of false for negative zero. From signum javadoc:
>Special Cases:
    If the argument is NaN, then the result is NaN.
    If the argument is positive zero or negative zero, then the result is the same as the argument. 

So if we want isZero false, this seems fine. But if we want isZero true, then we may need to add another `||` condition here for negative zero.

Either ways, we should add javadoc comment and test case for negative zero.

Reply:
I think -0 and +0 are fine. I'll add tests.